File: scripts/paper_identity.py
"""Strict cross-runtime paper identity contract (paper-identity-v1)."""

from __future__ import annotations

import hashlib
import ipaddress
import json
import re
from typing import Any, Mapping
from urllib.parse import urlsplit, urlunsplit

CONTRACT = "paper-identity-v1"
ARXIV_ID_RE = re.compile(r"^\d{4}\.\d{4,5}$")
SLUG_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
SCHEMES = frozenset(("icassp-arnumber", "openreview-forum-id", "conference-paper-id"))
SHA_RE = re.compile(r"^[a-f0-9]{64}$")


def _fail(message: str) -> None:
    raise ValueError(f"Invalid paper identity: {message}")


def _exact_object(value: Any, fields: tuple[str, ...], label: str) -> Mapping[str, Any]:
    if not isinstance(value, dict):
        _fail(f"{label} must be a plain object")
    if set(value) != set(fields):
        _fail(f"{label} has unknown or missing fields")
    return value


def _text(value: Any, label: str, *, allow_empty: bool = False, maximum: int = 4096) -> str:
    if not isinstance(value, str) or value != value.strip() or len(value) > maximum or (not allow_empty and not value):
        _fail(f"{label} must be a trimmed text value without controls")
    if any(ord(char) < 32 or ord(char) == 127 for char in value):
        _fail(f"{label} must be a trimmed text value without controls")
    return value


def _year(value: Any, label: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value < 1900 or value > 2100:
        _fail(f"{label} must be a supported four-digit year")
    return value


def validate_external_id(value: Any) -> dict[str, str]:
    raw = _exact_object(value, ("scheme", "value"), "externalId")
    if raw["scheme"] not in SCHEMES:
        _fail("externalId.scheme is unsupported")
    identifier = _text(raw["value"], "externalId.value", maximum=128)
    valid = (re.fullmatch(r"[1-9]\d*", identifier) if raw["scheme"] in {"icassp-arnumber", "conference-paper-id"}
             else re.fullmatch(r"[A-Za-z0-9_-]{6,128}", identifier))
    if not valid:
        _fail("externalId.value is invalid for its scheme")
    return {"scheme": raw["scheme"], "value": identifier}


def _public_dns_name(hostname: str | None) -> bool:
    if not hostname:
        return False
    host = hostname.lower()
    if host == "localhost" or host.endswith(".localhost") or ":" in host:
        return False
    try:
        ipaddress.ip_address(host)
        return False
    except ValueError:
        pass
    if len(host) > 253 or "." not in host:
        return False
    return all(re.fullmatch(r"[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?", label) for label in host.split("."))


def validate_official_url(value: Any, label: str) -> str:
    url = _text(value, label, maximum=2048)
    try:
        parsed = urlsplit(url)
    except ValueError:
        _fail(f"{label} is not a URL")
    try:
        port = parsed.port
    except ValueError:
        _fail(f"{label} is not a URL")
    if (parsed.scheme != "https" or not parsed.netloc or parsed.username or parsed.password or port
            or parsed.hostname != (parsed.hostname or "").lower()
            or parsed.query or parsed.fragment or not _public_dns_name(parsed.hostname)):
        _fail(f"{label} must be a canonical public HTTPS URL without credentials, port, query, or fragment")
    if (parsed.path == "/" or "//" in parsed.path or any(part in {".", ".."} or not re.fullmatch(r"[A-Za-z0-9._~-]+", part)
            for part in parsed.path.split("/")[1:])):
        _fail(f"{label} has an unsafe or non-canonical path")
    canonical = urlunsplit((parsed.scheme, parsed.netloc, parsed.path, "", ""))
    if canonical != url:
        _fail(f"{label} must use canonical URL spelling")
    return url


def validate_source(value: Any) -> dict[str, Any]:
    raw = _exact_object(value, ("status", "url"), "source")
    if raw["status"] == "unavailable":
        if raw["url"] is not None:
            _fail("source.url must be null when unavailable")
        return {"status": "unavailable", "url": None}
    if raw["status"] != "official":
        _fail("source.status must be official or unavailable")
    return {"status": "official", "url": validate_official_url(raw["url"], "source.url")}


def validate_citation(value: Any) -> dict[str, Any] | None:
    if value is None:
        return None
    raw = _exact_object(value, ("title", "authors", "venue", "year"), "citation")
    title = _text(raw["title"], "citation.title", maximum=2048)
    if not isinstance(raw["authors"], list) or len(raw["authors"]) > 100:
        _fail("citation.authors must be an array of at most 100 names")
    authors = [_text(author, f"citation.authors[{index}]", maximum=512) for index, author in enumerate(raw["authors"])]
    if len(set(authors)) != len(authors):
        _fail("citation.authors must not contain duplicate names")
    venue = None if raw["venue"] is None else _text(raw["venue"], "citation.venue", maximum=512)
    year = None if raw["year"] is None else _year(raw["year"], "citation.year")
    return {"title": title, "authors": authors, "venue": venue, "year": year}


def canonical_conference_id(conference: Mapping[str, Any], external_id: Mapping[str, str]) -> str:
    return f"conference:{conference['slug']}:{conference['year']}:{external_id['scheme']}:{external_id['value']}"


def conference_coordinates(conference: Any) -> dict[str, Any]:
    raw = _exact_object(conference, ("id", "year"), "conference coordinates")
    year = _year(raw["year"], "conference coordinates.year")
    suffix = f"-{year}"
    if not isinstance(raw["id"], str) or not raw["id"].endswith(suffix):
        _fail("conference coordinates.id must end with its exact year")
    slug = raw["id"][:-len(suffix)]
    if not SLUG_RE.fullmatch(slug):
        _fail("conference coordinates.id must contain a normalized slug")
    return {"slug": slug, "year": year}


def canonical_conference_paper_id(conference: Any, source_identity: Any) -> str:
    raw = _exact_object(source_identity, ("type", "value"), "conference source identity")
    external_id = validate_external_id({"scheme": raw["type"], "value": raw["value"]})
    return canonical_conference_id(conference_coordinates(conference), external_id)


def assert_canonical_conference_paper_id(value: Any, conference: Any, source_identity: Any) -> str:
    expected = canonical_conference_paper_id(conference, source_identity)
    if value != expected:
        _fail(f"conference paperId must be {expected}")
    return expected


def normalize_identity(value: Any) -> dict[str, Any]:
    raw = _exact_object(value, ("contract", "kind", "canonicalId", "arxivId", "conference", "externalId", "source", "citation"), "paper identity")
    if raw["contract"] != CONTRACT:
        _fail(f"contract must be {CONTRACT}")
    if raw["kind"] not in {"arxiv", "conference"}:
        _fail("kind must be arxiv or conference")
    source = validate_source(raw["source"])
    citation = validate_citation(raw["citation"])
    if raw["kind"] == "arxiv":
        if not isinstance(raw["arxivId"], str) or not ARXIV_ID_RE.fullmatch(raw["arxivId"]):
            _fail("arxivId is invalid")
        if raw["conference"] is not None or raw["externalId"] is not None:
            _fail("arxiv identity must not contain conference fields")
        canonical_id = f"arxiv:{raw['arxivId']}"
        if raw["canonicalId"] != canonical_id:
            _fail("canonicalId does not bind arxivId")
        if source["status"] == "official" and source["url"] != f"https://arxiv.org/abs/{raw['arxivId']}":
            _fail("official arxiv source.url must bind its exact canonical abs URL")
        return {"contract": CONTRACT, "kind": "arxiv", "canonicalId": canonical_id, "arxivId": raw["arxivId"],
                "conference": None, "externalId": None, "source": source, "citation": citation}
    if raw["arxivId"] is not None:
        _fail("conference identity must not contain arxivId")
    conference_raw = _exact_object(raw["conference"], ("slug", "year"), "conference")
    if not isinstance(conference_raw["slug"], str) or not SLUG_RE.fullmatch(conference_raw["slug"]):
        _fail("conference.slug must be a normalized slug")
    conference = {"slug": conference_raw["slug"], "year": _year(conference_raw["year"], "conference.year")}
    external_id = validate_external_id(raw["externalId"])
    canonical_id = canonical_conference_id(conference, external_id)
    if raw["canonicalId"] != canonical_id:
        _fail("canonicalId does not bind conference, year, scheme, and value")
    return {"contract": CONTRACT, "kind": "conference", "canonicalId": canonical_id, "arxivId": None,
            "conference": conference, "externalId": external_id, "source": source, "citation": citation}


def stable_json(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"), allow_nan=False)


def sha256(value: str | bytes) -> str:
    return hashlib.sha256(value.encode("utf-8") if isinstance(value, str) else value).hexdigest()


def stable_sha256(value: Any) -> str:
    return sha256(stable_json(value))


def identity_payload(value: Any) -> dict[str, Any]:
    normalized = normalize_identity(value)
    if normalized["kind"] == "arxiv":
        return {"contract": CONTRACT, "kind": "arxiv", "canonicalId": normalized["canonicalId"], "arxivId": normalized["arxivId"]}
    return {"contract": CONTRACT, "kind": "conference", "canonicalId": normalized["canonicalId"],
            "conference": normalized["conference"], "externalId": normalized["externalId"]}


def identity_sha256(value: Any) -> str:
    return stable_sha256(identity_payload(value))


def record_sha256(value: Any) -> str:
    return stable_sha256(normalize_identity(value))


def is_sha256(value: Any) -> bool:
    return bool(SHA_RE.fullmatch(str(value or "")))


__all__ = ["CONTRACT", "ARXIV_ID_RE", "SCHEMES", "canonical_conference_id", "conference_coordinates",
           "canonical_conference_paper_id", "assert_canonical_conference_paper_id", "validate_external_id",
           "validate_official_url", "validate_source", "validate_citation", "normalize_identity", "identity_payload",
           "stable_json", "sha256", "stable_sha256", "identity_sha256", "record_sha256", "is_sha256"]


if __name__ == '__main__':
    # Every top-level Python script has the same external-runtime boundary,
    # even when this file normally serves only as an importable shared module.
    from runtime_guard import require_external_runtime
    require_external_runtime('paper_identity.py')

File: scripts/lib/paper-identity.js
'use strict';

// Cross-runtime identity record for papers that may be sourced from arXiv or
// conference proceedings.  This deliberately does not import the historic
// arXiv helpers: consumers must opt into this v1 contract instead of silently
// changing legacy daily-digest identity behaviour.

const crypto = require('node:crypto');

const CONTRACT = 'paper-identity-v1';
const ARXIV_ID_RE = /^\d{4}\.\d{4,5}$/;
const SLUG_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const SCHEMES = new Set(['icassp-arnumber', 'openreview-forum-id', 'conference-paper-id']);
const SHA_RE = /^[a-f0-9]{64}$/;

function fail(message) { throw new Error(`Invalid paper identity: ${message}`); }

function assertPlainObject(value, label) {
    if (!value || typeof value !== 'object' || Array.isArray(value)
        || ![Object.prototype, null].includes(Object.getPrototypeOf(value))) {
        fail(`${label} must be a plain object`);
    }
}

function assertExactFields(value, fields, label) {
    assertPlainObject(value, label);
    const actual = Object.keys(value).sort();
    const expected = [...fields].sort();
    if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index])) {
        fail(`${label} has unknown or missing fields`);
    }
}

function assertText(value, label, { allowEmpty = false, max = 4096 } = {}) {
    if (typeof value !== 'string' || value !== value.trim() || value.length > max
        || (!allowEmpty && !value) || /[\u0000-\u001f\u007f]/u.test(value)) {
        fail(`${label} must be a trimmed text value without controls`);
    }
    return value;
}

function assertYear(value, label) {
    if (!Number.isInteger(value) || value < 1900 || value > 2100) fail(`${label} must be a supported four-digit year`);
    return value;
}

function validateExternalId(value) {
    assertExactFields(value, ['scheme', 'value'], 'externalId');
    if (!SCHEMES.has(value.scheme)) fail('externalId.scheme is unsupported');
    assertText(value.value, 'externalId.value', { max: 128 });
    const valid = (value.scheme === 'icassp-arnumber' || value.scheme === 'conference-paper-id')
        ? /^[1-9]\d*$/.test(value.value)
        : /^[A-Za-z0-9_-]{6,128}$/.test(value.value);
    if (!valid) fail('externalId.value is invalid for its scheme');
    return { scheme: value.scheme, value: value.value };
}

function isPublicDnsName(hostname) {
    // The source URL is a display/provenance pointer, not a network admission
    // decision.  Still reject literals and special-use local names here so an
    // identity record cannot smuggle a localhost/private URL into a renderer.
    if (!hostname || hostname === 'localhost' || hostname.endsWith('.localhost')
        || hostname.includes(':') || /^\d{1,3}(?:\.\d{1,3}){3}$/.test(hostname)) return false;
    if (hostname.length > 253 || !hostname.includes('.')) return false;
    return hostname.split('.').every(label => /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/i.test(label));
}

function validateOfficialUrl(value, label) {
    assertText(value, label, { max: 2048 });
    let parsed;
    try { parsed = new URL(value); } catch { fail(`${label} is not a URL`); }
    if (parsed.protocol !== 'https:' || parsed.username || parsed.password || parsed.port
        || parsed.search || parsed.hash || !isPublicDnsName(parsed.hostname)) {
        fail(`${label} must be a canonical public HTTPS URL without credentials, port, query, or fragment`);
    }
    // Require a non-root, portable, unambiguous path.  Checking both the raw
    // spelling and URL form prevents URL() from normalising a traversal away.
    const rawPath = value.slice(`https://${parsed.host}`.length);
    if (parsed.pathname === '/' || rawPath !== parsed.pathname || parsed.pathname.includes('//')
        || parsed.pathname.split('/').slice(1).some(part => !/^[A-Za-z0-9._~-]+$/.test(part))) {
        fail(`${label} has an unsafe or non-canonical path`);
    }
    if (parsed.toString() !== value) fail(`${label} must use canonical URL spelling`);
    return value;
}

function validateSource(value) {
    assertExactFields(value, ['status', 'url'], 'source');
    if (value.status === 'unavailable') {
        if (value.url !== null) fail('source.url must be null when unavailable');
        return { status: 'unavailable', url: null };
    }
    if (value.status !== 'official') fail('source.status must be official or unavailable');
    return { status: 'official', url: validateOfficialUrl(value.url, 'source.url') };
}

function validateCitation(value) {
    if (value === null) return null;
    assertExactFields(value, ['title', 'authors', 'venue', 'year'], 'citation');
    assertText(value.title, 'citation.title', { max: 2048 });
    if (!Array.isArray(value.authors) || value.authors.length > 100) fail('citation.authors must be an array of at most 100 names');
    const authors = value.authors.map((author, index) => assertText(author, `citation.authors[${index}]`, { max: 512 }));
    if (new Set(authors).size !== authors.length) fail('citation.authors must not contain duplicate names');
    if (value.venue !== null) assertText(value.venue, 'citation.venue', { max: 512 });
    if (value.year !== null) assertYear(value.year, 'citation.year');
    return { title: value.title, authors, venue: value.venue, year: value.year };
}

function canonicalConferenceId(conference, externalId) {
    return `conference:${conference.slug}:${conference.year}:${externalId.scheme}:${externalId.value}`;
}

function normalizeIdentity(value) {
    assertExactFields(value, ['contract', 'kind', 'canonicalId', 'arxivId', 'conference', 'externalId', 'source', 'citation'], 'paper identity');
    if (value.contract !== CONTRACT) fail(`contract must be ${CONTRACT}`);
    if (value.kind !== 'arxiv' && value.kind !== 'conference') fail('kind must be arxiv or conference');
    const source = validateSource(value.source);
    const citation = validateCitation(value.citation);

    if (value.kind === 'arxiv') {
        if (!ARXIV_ID_RE.test(String(value.arxivId || ''))) fail('arxivId is invalid');
        if (value.conference !== null || value.externalId !== null) fail('arxiv identity must not contain conference fields');
        const canonicalId = `arxiv:${value.arxivId}`;
        if (value.canonicalId !== canonicalId) fail('canonicalId does not bind arxivId');
        if (source.status === 'official' && source.url !== `https://arxiv.org/abs/${value.arxivId}`) {
            fail('official arxiv source.url must bind its exact canonical abs URL');
        }
        return { contract: CONTRACT, kind: 'arxiv', canonicalId, arxivId: value.arxivId,
            conference: null, externalId: null, source, citation };
    }

    if (value.arxivId !== null) fail('conference identity must not contain arxivId');
    assertExactFields(value.conference, ['slug', 'year'], 'conference');
    if (!SLUG_RE.test(String(value.conference.slug || ''))) fail('conference.slug must be a normalized slug');
    const conference = { slug: value.conference.slug, year: assertYear(value.conference.year, 'conference.year') };
    const externalId = validateExternalId(value.externalId);
    const canonicalId = canonicalConferenceId(conference, externalId);
    if (value.canonicalId !== canonicalId) fail('canonicalId does not bind conference, year, scheme, and value');
    return { contract: CONTRACT, kind: 'conference', canonicalId, arxivId: null,
        conference, externalId, source, citation };
}

function canonicalize(value) {
    if (Array.isArray(value)) return value.map(canonicalize);
    if (value && typeof value === 'object') {
        return Object.fromEntries(Object.keys(value).sort().map(key => [key, canonicalize(value[key])]));
    }
    return value;
}

function stableJson(value) { return JSON.stringify(canonicalize(value)); }
function sha256(value) { return crypto.createHash('sha256').update(value).digest('hex'); }
function stableSha256(value) { return sha256(stableJson(value)); }
function identityPayload(value) {
    const normalized = normalizeIdentity(value);
    return normalized.kind === 'arxiv'
        ? { contract: CONTRACT, kind: 'arxiv', canonicalId: normalized.canonicalId, arxivId: normalized.arxivId }
        : { contract: CONTRACT, kind: 'conference', canonicalId: normalized.canonicalId,
            conference: normalized.conference, externalId: normalized.externalId };
}
function identitySha256(value) { return stableSha256(identityPayload(value)); }
function recordSha256(value) { return stableSha256(normalizeIdentity(value)); }
function isSha256(value) { return SHA_RE.test(String(value || '')); }

module.exports = {
    CONTRACT, ARXIV_ID_RE, SCHEMES, canonicalConferenceId, validateExternalId,
    validateOfficialUrl, validateSource, validateCitation, normalizeIdentity,
    identityPayload, stableJson, sha256, stableSha256, identitySha256, recordSha256, isSha256
};
